File: Namrata-Sanger/301. Remove Invalid Parentheses.js
/*
Day 32
301. Remove Invalid Parentheses
Link: https://leetcode.com/problems/remove-invalid-parentheses/submissions/
Level : Hard
*/

var removeInvalidParentheses = function(s) {
  const ans = [];

  const isValid = s => {
    let cnt = 0;
    for (const ch of s) {
      if (ch == "(") cnt += 1;
      if (ch == ")") cnt -= 1;
      if (cnt < 0) return false;
    }
    return cnt == 0;
  };

  const dfs = (s, start, l, r) => {
    if (!l && !r) {
      if (isValid(s)) ans.push(s);
      return;
    }
    for (let i = start; i != s.length; i += 1) {
      if (i != start && s[i] == s[i - 1]) continue;
      if (r) {
        if (s[i] != ")") continue;
        dfs(s.substr(0, i) + s.substr(i + 1), i, l, r - 1);
      } else {
        if (s[i] != "(") continue;
        dfs(s.substr(0, i) + s.substr(i + 1), i, l - 1, r);
      }
    }
  };

  let l = 0, r = 0;
  for (const ch of s) {
    l += ch == "(";
    if (l == 0) {
      r += ch == ")";
    } else {
      l -= ch == ")";
    }
  }
  dfs(s, 0, l, r);
  return ans;
}

File: Vindeep-Chaudhari/301_Remove_Invalid_Parentheses.cpp
/**
 * Problem Link - https://leetcode.com/problems/remove-invalid-parentheses/
 */

class Solution {
public:
    bool isValidP(string s) {
        int sum = 0;
        for(char &c : s) {
            if(c == '(') ++sum;
            else if(c == ')') --sum;
            if(sum < 0) return false;
        }
        return sum == 0;
    }
    vector<string> removeInvalidParentheses(string s) {
        int open = 0, close = 0;
        for(char &c : s) {
            open += c == '(';
            if (open == 0) {
                close += c == ')';
            } else {
                open -= c == ')';
            }
        }
        vector<string> ans;
        dfs(s, 0, open, close, ans);
        return ans;
    }
    void dfs(string s, int beg, int open, int close, vector<string> &ans) {
        if(open == 0 && close == 0) {
            if(isValidP(s))
                ans.push_back(s);
        } else {
            for(int i = beg; i < s.size(); ++i) {
                string tmp = s;
                if(open > 0 && tmp[i] == '(') {
                    if(i == beg || tmp[i] != tmp[i - 1]) {
                        tmp.erase(i, 1);
                        dfs(tmp, i, open - 1, close, ans);
                    }
                }
                if(close > 0 && tmp[i] == ')') {
                    if(i == beg || tmp[i] != tmp[i - 1]) {
                        tmp.erase(i, 1);
                        dfs(tmp, i, open, close - 1, ans);
                    }
                }
            }
        }
    }
};
